src/Lecture2.hs

```diff
@@ -243,53 +371,78 @@ Add (Var "x") (Lit 1)
-It returns either a successful evaluation result or an error.
--}
+-- | Having all this set up, we can finally implement an evaluation function.
+-- It returns either a successful evaluation result or an error.
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
-
-{- | Compilers also perform optimizations! One of the most common
-optimizations is "Constant Folding". It performs arithmetic operations
-on all constants known during compile time. This way you can write
-more verbose and clear code that works as efficient as its shorter
-version.
-
-For example, if you have an expression:
-
-x + 10 + y + 15 + 20
-
-The result of constant folding can be:
-
-x + y + 45
-
-It also can be:
-
-x + 45 + y
-
-Write a function that takes and expression and performs "Constant
-Folding" optimization on the given expression.
--}
+eval _ (Lit i) = Right i
+eval vs (Var v) = case lookup v vs of
+  Just val -> Right val
+  Nothing -> Left (VariableNotFound v)
+eval vs (Add e1 e2) = case (eval vs e1, eval vs e2) of
+  (Left err, _) -> Left err
+  (_, Left err) -> Left err
+  (Right i, Right j) -> Right (i + j)
+
+-- | Compilers also perform optimizations! One of the most common
+-- optimizations is "Constant Folding". It performs arithmetic operations
+-- on all constants known during compile time. This way you can write
+-- more verbose and clear code that works as efficient as its shorter
+-- version.
+--
+-- For example, if you have an expression:
+--
+-- x + 10 + y + 15 + 20
+--
+-- The result of constant folding can be:
+--
+-- x + y + 45
+--
+-- It also can be:
+--
+-- x + 45 + y
+--
+-- Write a function that takes and expression and performs "Constant
+-- Folding" optimization on the given expression.
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding e = case e of
+  Lit i -> Lit i
+  Var x -> Var x
+  Add (Lit i) (Lit j) -> Lit (i + j)
+  Add (Lit 0) e2 -> constantFolding e2
+  Add e1 (Lit 0) -> constantFolding e1
+  -- When there are "recursive" add expressions containing literals,
+  -- we group literals together.
+  -- Form 1: 10 + (x + 20)
+  Add (Lit i) e2 -> case e2 of
+    Add expr (Lit j) -> constantFolding (Add expr (Lit (i + j)))
+    Add (Lit j) expr -> constantFolding (Add expr (Lit (i + j)))
+    _ -> Add (constantFolding e2) (Lit i)
+  Add e1 (Lit i) -> case e1 of
+    Add expr (Lit j) -> constantFolding (Add expr (Lit (i + j)))
+    Add (Lit j) expr -> constantFolding (Add expr (Lit (i + j)))
+    _ -> Add (constantFolding e1) (Lit i)
+  -- Form 2: (x + 10) + (y + 20)
+  Add (Add (Lit i) e1) (Add (Lit j) e2) -> constantFolding (Add (Add e1 e2) (Lit (i + j)))
+  Add (Add e1 (Lit i)) (Add e2 (Lit j)) -> constantFolding (Add (Add e1 e2) (Lit (i + j)))
+  Add (Add (Lit i) e1) (Add e2 (Lit j)) -> constantFolding (Add (Add e1 e2) (Lit (i + j)))
+  Add (Add e1 (Lit i)) (Add (Lit j) e2) -> constantFolding (Add (Add e1 e2) (Lit (i + j)))
+  Add e1 e2 -> Add (constantFolding e1) (constantFolding e2)
```

Comment:
I see that you've implemented a nice pattern matching here 👍🏻 

An alternative solution would be to write a helper function that traverses the expression tree and combines all the results into the list of all variable names and addition of all constants, e.g. into a type like `([String], Int)`. After that, it's trivial to reconstruct the optimal expression by pattern matching on the about structure.

If you want to go the route of rebalancing the expression tree, you can maintain a form of e.g.

```haskell
Lit n `Add` (Var x `Add` (Var y `Add` (Var z `Add` ... )))
```

So the literal is the left upmost son of `Add`. So you kinda pull all constants up to the root.

That way you'll have much fewer cases to handle.

Reply:
At first, I tried using an helper function but got lost in the middle of the implementation. That's why I went the pattern matching route but I found that a bit verbose and "brute force" even though it works fine. 

Now, I see how it could be done thanks to your suggestion. Thanks a lot !

I'll try to add this new implementation.